doc(polys): explain the domain system

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

fixes #19583 
fixes #19585

#### Brief description of what is fixed or changed

Adds tutorial-style docs aimed at developers to introduce the domain system

#### Other comments

I'm not sure where this really belongs. We need good places in the docs to separate the different information for different audiences...

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

oscarbenjamin committed:
doc(polys): explain the domain system

oscarbenjamin committed:
doc(polys): explain the domain system

Add docstring for DomainElement.parent()

gschintgen commented:
Hi, I really like the documentation you're proposing in this PR! These explanations are definitely needed.
I'd like to make a few minor suggestions:
- It'd be nice to mention other available domains: ring of residues modulo an integer, `ZZ_I`, `QQ_I` (#19583). Even without detailed explanations (e.g. just the basic syntax) this could be valuable for newcomers.
- A link/reference to these new explanations from the "Basic concepts" section of the polys docs would be helpful. Currently the `polys` documentation is a bit vague on the available domains and their recommended syntax.
- I suppose `domain='QQ<sqrt(2)>'` as shown in the examples of the basic concepts is equivalent to `QQ.algebraic_field(sqrt(2))` just that the former uses a special string parser?

oscarbenjamin committed:
doc(polys): explain the domain system

Move domains docs to a separate page and add background

oscarbenjamin commented:
> I'd like to make a few minor suggestions:

Those are good suggestions. I'll do that.

> * I suppose `domain='QQ<sqrt(2)>'` as shown in the examples of the basic concepts is equivalent to `QQ.algebraic_field(sqrt(2))` just that the former uses a special string parser?

Yes, exactly:

```
In [3]: Poly(x**2 + sqrt(2), extension=True)
Out[3]: Poly(x**2 + sqrt(2), x, domain='QQ<sqrt(2)>')

In [4]: Poly(x**2 + sqrt(2), extension=True).domain
Out[4]: QQ<sqrt(2)>

In [5]: Poly(x**2 + sqrt(2), extension=True).domain == QQ.algebraic_field(sqrt(2))
Out[5]: True
```

oscarbenjamin committed:
doc(polys): explain the domain system

oscarbenjamin committed:
doc(polys): explain the domain system

oscarbenjamin committed:
doc(polys): explain the domain system

oscarbenjamin committed:
doc(polys): explain the domain system

oscarbenjamin committed:
doc(polys): explain the domain system

oscarbenjamin committed:
doc(polys): explain the domain system

oscarbenjamin committed:
doc(polys): explain the domain system

oscarbenjamin committed:
doc(polys): explain the domain system

oscarbenjamin committed:
doc(polys): explain the domain system

oscarbenjamin committed:
doc(polys): explain the domain system

oscarbenjamin committed:
doc(polys): explain the domain system

oscarbenjamin committed:
doc(polys): explain the domain system

Move docs for domain classes from internals.rst into a new
domainsref.rst. Rename domains.rst to domainsintro.rst.

oscarbenjamin committed:
doc(polys): explain the domain system

Reorder headings and sections in domainsref.rst.

oscarbenjamin committed:
doc(polys): explain the domain system

Make clickable references to ZZ and QQ and add PythonRational to docs

oscarbenjamin committed:
doc(polys): explain the domain system

Add missing classes to the docs and make references clickable in
domainsintro.rst.

oscarbenjamin committed:
doc(polys): explain the domain system

Replace references to Basic with Expr

oscarbenjamin committed:
doc(polys): explain the domain system

Add docs for ZZ and QQ

oscarbenjamin committed:
doc(polys): explain the domain system

oscarbenjamin committed:
doc(polys): explain the domain system

Extend the Domain docstring

oscarbenjamin committed:
doc(polys): explain the domain system

Add docs for methods and attributes of Domain.

oscarbenjamin committed:
doc(polys): explain the domain system

Add docs for K.to_sympy and K.from_sympy.

oscarbenjamin committed:
doc(polys): explain the domain system

Add docs for ZZ_I

oscarbenjamin committed:
doc(polys): explain the domain system

Add docs for QQ_I

oscarbenjamin committed:
doc(polys): explain the domain system

Add docs for GF(p)

oscarbenjamin committed:
doc(polys): explain the domain system

Add examples of using the domains for QQ_I and GF(p)

oscarbenjamin commented:
This is not yet finished but it would be great to have a review @gschintgen @jksuom @sylee957 @asmeurer @Soumi7 

This adds a load of docs for things that are basically undocumented at the moment.

oscarbenjamin commented on sympy/polys/domains/domain.py:
```diff
@@ -14,16 +14,301 @@
+
+    """
+
+    #: The type (class) of the elements of this :py:class:`~.Domain`:
+    #:
+    #: >>> from sympy import ZZ, QQ, Symbol
+    #: >>> ZZ.dtype
+    #: <class 'int'>
+    #: >>> z = ZZ(2)
+    #: >>> z
+    #: 2
+    #: >>> type(z)
+    #: <class 'int'>
+    #: >>> type(z) == ZZ.dtype
+    #: True
+    #:
+    #: Every domain has an associated **dtype** ("datatype") which is the
+    #: class of the associated domain elements.
+    #:
+    #: See also
+    #: ========
+    #:
+    #: of_type
+    #:
+    dtype = None        # type: Optional[Type]
```
I like this style of documentation fo class attributes because it shows in the code and the docs but I don't think it's properly doctested. Is there a better/recommended way to document class attributes?

oscarbenjamin committed:
doc(polys): explain the domain system

oscarbenjamin committed:
doc(polys): explain the domain system

Add docs for QQ<a>

oscarbenjamin committed:
doc(polys): explain the domain system

Add docs for construct_domain

oscarbenjamin commented:
I'm about to become extremely busy at work so I've gone as far as I can with this.

Please review!

Soumi7 reviewed (commented):
Looks good to me.

sylee957 commented:
I'd try to dedent every examples since this is consistent with most of the sympy.
Although indentation is recognized by HTML build such that it's shifted a bit rightwards, I think that it's better not to have unnecessary indentation due to PEP-8 line limits.

oscarbenjamin committed:
doc(polys): explain the domain system

Dedent examples sections in docstrings.

gschintgen commented:
I'd say this PR fixes #19583 and #19585.

gschintgen commented on doc/src/modules/polys/domainsintro.rst:
```diff
@@ -0,0 +1,1548 @@
+What are the domains?
+=====================
+
+For most users the domains are only really noticable in the printed output of
```
noticeable

gschintgen commented on doc/src/modules/polys/domainsintro.rst:
```diff
@@ -0,0 +1,1548 @@
+  >>> z**2
+  (-3 + 4*I)
+
+Note the constrast with the way this calculation works in the tree
```
contrast

gschintgen commented on doc/src/modules/polys/domainsintro.rst:
```diff
@@ -0,0 +1,1548 @@
+efficient than the tree representation used by :py:class:`~.Expr`. In
+:ref:`EX` the internal representation is :py:class:`~.Expr` so it is clearly
+not more efficient. The purpose of the :ref:`EX` domain is to be able to wrap
+up arbitrary expressions in an interace that is consistent with the other
```
interface

gschintgen commented on doc/src/modules/polys/domainsintro.rst:
```diff
@@ -0,0 +1,1548 @@
+  True
+
+The :py:meth:`~.Domain.convert` method can be called without specifying the
+source domain as the second ergument e.g.::
```
argument

gschintgen commented on doc/src/modules/polys/domainsref.rst:
```diff
@@ -0,0 +1,338 @@
+===================================
+
+This page lists the reference documentation for the domains in the polys
+module. For a general introduction to the polts module it is recommended to
```
polys

gschintgen commented on doc/src/modules/polys/domainsref.rst:
```diff
@@ -0,0 +1,338 @@
+
+This page lists the reference documentation for the domains in the polys
+module. For a general introduction to the polts module it is recommended to
+read :ref:`polys-basics` instead. For an introductoray explanation of the
```
introductory

gschintgen reviewed (commented):
I read through an earlier (incomplete) version of this changeset and found it to be very well written. It's also a much needed addition. Thanks and +1. (I did not render to html and re-read every line though.)

oscarbenjamin committed:
doc(polys): explain the domain system

Fix typos

oscarbenjamin commented:
> I'd try to dedent every examples

I've dedented them now.

> I'd say this PR fixes #19583 and #19585.

I forgot about those. I've added the "fixes" for them to the OP

Thanks for the reviews @gschintgen and @sylee957

oscarbenjamin committed:
Merge branch 'master' into pr_domaindocs

oscarbenjamin commented:
Any more reviews?

sylee957 commented:
This is the first time introducing docstring comment `#:` in sympy. But I'm not sure if there are other ways around using ordinary docstring like https://stackoverflow.com/questions/56693832/should-sphinx-be-able-to-document-instance-attributes-in-a-class.
I'm not sure if that can help doctests, but I'd prefer to use ordinary docstrings than comments if possible.

oscarbenjamin commented:
> I'm not sure if that can help doctests, but I'd prefer to use ordinary docstrings than comments if possible.

We can use those. I think that to make either those or the comments work as doctests the doctest runner needs to parse the code to extract the examples because neither the doc comments or the attribute "docstrings" that you refer to are actually available as `__doc__` anywhere. I don't know if other doctest runners can do this but I doubt that sympy's own internal one can.

There is another possibility which is to have an `attributes` section in main class docstring but when I tried that it didn't come out very nicely: it made a small table with not much space to discuss each attribute and have doctested examples. I don't know if there's a way to improve that.

oscarbenjamin committed:
doc(polys): use attribute docstrings rather than doc-comments

oscarbenjamin committed:
doc(polys): add ExpressionDomain.Expression to the docs

oscarbenjamin committed:
Merge branch 'master' into pr_domaindocs

oscarbenjamin committed:
doc(polys): use rawstrings for docstrings with latex

oscarbenjamin committed:
doc(polys): use rawstrings for docstrings with latex

oscarbenjamin committed:
doc(polys): use rawstrings for docstrings with latex

oscarbenjamin committed:
doc(polys): skip doctests that differ with gmpy2

oscarbenjamin committed:
doc(polys): skip doctests that differ with gmpy2

oscarbenjamin committed:
doc(polys): format operators as code (not latex)

oscarbenjamin committed:
doc(polys): small tweaks in domain docs

oscarbenjamin commented:
> I'm not sure if that can help doctests, but I'd prefer to use ordinary docstrings than comments if possible.

I changed them to attribute "docstrings" rather than doc-comments.
